File: app/src/main/java/com/o4x/musical/ui/activities/MainActivity.java
package com.o4x.musical.ui.activities;

import android.annotation.SuppressLint;
import android.content.Intent;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.graphics.drawable.GradientDrawable;
import android.graphics.drawable.LayerDrawable;
import android.graphics.drawable.StateListDrawable;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.provider.MediaStore;
import android.util.Log;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.drawerlayout.widget.DrawerLayout;
import androidx.fragment.app.Fragment;

import com.google.android.material.navigation.NavigationView;
import com.o4x.musical.App;
import com.o4x.musical.R;
import com.o4x.musical.equalizer.EqualizerFragment;
import com.o4x.musical.helper.MusicPlayerRemote;
import com.o4x.musical.helper.SearchQueryHelper;
import com.o4x.musical.imageloader.universalil.UniversalIL;
import com.o4x.musical.loader.AlbumLoader;
import com.o4x.musical.loader.ArtistLoader;
import com.o4x.musical.loader.PlaylistSongLoader;
import com.o4x.musical.model.Song;
import com.o4x.musical.service.MusicService;
import com.o4x.musical.ui.activities.base.AbsMusicPanelActivity;
import com.o4x.musical.ui.activities.intro.AppIntroActivity;
import com.o4x.musical.ui.dialogs.ChangelogDialog;
import com.o4x.musical.ui.dialogs.ScanMediaFolderChooserDialog;
import com.o4x.musical.ui.fragments.mainactivity.folders.FoldersFragment;
import com.o4x.musical.ui.fragments.mainactivity.home.HomeFragment;
import com.o4x.musical.ui.fragments.mainactivity.library.LibraryFragment;
import com.o4x.musical.ui.fragments.mainactivity.queue.QueueFragment;
import com.o4x.musical.util.PreferenceUtil;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import code.name.monkey.appthemehelper.ThemeStore;
import code.name.monkey.appthemehelper.util.ATHUtil;
import code.name.monkey.appthemehelper.util.ColorUtil;
import code.name.monkey.appthemehelper.util.NavigationViewUtil;

public class MainActivity extends AbsMusicPanelActivity {

    public static final String TAG = MainActivity.class.getSimpleName();
    public static final int APP_INTRO_REQUEST = 100;


    @BindView(R.id.navigation_view)
    NavigationView navigationView;
    @BindView(R.id.drawer_layout)
    DrawerLayout drawerLayout;

    @Nullable
    MainActivityFragmentCallbacks currentFragment;

    private boolean blockRequestPermissions;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setDrawUnderBar();
        ButterKnife.bind(this);

        setUpDrawerLayout();

        if (savedInstanceState == null) {
            setMusicChooser(PreferenceUtil.getLastMusicChooser());
        } else {
            restoreCurrentFragment();
        }

        if (!checkShowIntro()) {
            showChangelog();
        }

        App.setOnProVersionChangedListener(() -> {
            // called if the cached value was outdated (should be a rare event)
            checkSetUpPro();
            if (!App.isProVersion() && PreferenceUtil.getLastMusicChooser() == R.id.nav_folders) {
                setMusicChooser(R.id.nav_folders); // shows the purchase activity and switches to LIBRARY
            }
        });

        UniversalIL.initImageLoader(this);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        App.setOnProVersionChangedListener(null);
    }

    public void setMusicChooser(int id) {
        if (!App.isProVersion() && id == R.id.nav_folders) {
            Toast.makeText(this, R.string.folder_view_is_a_pro_feature, Toast.LENGTH_LONG).show();
            startActivity(new Intent(this, PurchaseActivity.class));
            id = R.id.nav_library;
        }

        PreferenceUtil.setLastMusicChooser(id);
        navigationView.setCheckedItem(id);
        switch (id) {
            case R.id.nav_home:
                setCurrentFragment(HomeFragment.newInstance());
                break;
            case R.id.nav_queue:
                setCurrentFragment(QueueFragment.newInstance());
                break;
            case R.id.nav_library:
                setCurrentFragment(LibraryFragment.newInstance());
                break;
            case R.id.nav_folders:
                setCurrentFragment(FoldersFragment.newInstance(this));
                break;
            case R.id.nav_eq:
                setCurrentFragment(
                        EqualizerFragment.newBuilder()
                        .setthemeColor(Color.parseColor("#4caf50"))
                        .setAudioSessionId(MusicPlayerRemote.getAudioSessionId())
                        .build()
                );
                break;
        }
    }

    private void setCurrentFragment(@SuppressWarnings("NullableProblems") Fragment fragment) {
        getSupportFragmentManager().beginTransaction().replace(R.id.fragment_container, fragment, null).commit();
        currentFragment = (MainActivityFragmentCallbacks) fragment;
    }

    private void restoreCurrentFragment() {
        currentFragment = (MainActivityFragmentCallbacks) getSupportFragmentManager().findFragmentById(R.id.fragment_container);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == APP_INTRO_REQUEST) {
            blockRequestPermissions = false;
            if (!hasPermissions()) {
                requestPermissions();
            }
        }
    }

    @Override
    protected void requestPermissions() {
        if (!blockRequestPermissions) super.requestPermissions();
    }

    @Override
    protected View createContentView() {
        @SuppressLint("InflateParams")
        View contentView = getLayoutInflater().inflate(R.layout.activity_main_drawer_layout, null);
        ViewGroup drawerContent = contentView.findViewById(R.id.drawer_content_container);
        drawerContent.addView(wrapSlidingMusicPanel(R.layout.activity_main_content));

        // To apply WindowInsets only for navigation view, not content and it's very important.
        contentView.setOnApplyWindowInsetsListener((view, windowInsets) -> {
            view.findViewById(R.id.navigation_view).onApplyWindowInsets(windowInsets);
            view.findViewById(R.id.drawer_content_container).setPadding(
                    0,0,0,windowInsets.getSystemWindowInsetBottom());
            return windowInsets;
        });

        return contentView;
    }



    private void setUpNavigationView() {
        int themeColor = ThemeStore.Companion.themeColor(this);
        NavigationViewUtil.INSTANCE.setItemIconColors(navigationView, ATHUtil.INSTANCE.resolveColor(this, R.attr.iconColor, ThemeStore.Companion.textColorSecondary(this)), themeColor);
        NavigationViewUtil.INSTANCE.setItemTextColors(navigationView, ThemeStore.Companion.textColorPrimary(this), themeColor);

        StateListDrawable stateListDrawable = (StateListDrawable) navigationView.getItemBackground();
        LayerDrawable layerDrawable = (LayerDrawable) stateListDrawable.getStateDrawable(0);
        GradientDrawable rectangle = (GradientDrawable) layerDrawable.findDrawableByLayerId(R.id.rectangle);
        GradientDrawable rectangleRadius = (GradientDrawable) layerDrawable.findDrawableByLayerId(R.id.rectangle_radius);

        rectangle.setColor(themeColor);
        rectangleRadius.setColor(ColorUtil.INSTANCE.withAlpha(ThemeStore.Companion.textColorSecondary(this), 0.1f));

        checkSetUpPro();
        navigationView.setNavigationItemSelectedListener(menuItem -> {
            drawerLayout.closeDrawers();
            switch (menuItem.getItemId()) {
                case R.id.nav_home:
                    new Handler().postDelayed(() -> setMusicChooser(R.id.nav_home), 200);
                    break;
                case R.id.nav_queue:
                    new Handler().postDelayed(() -> setMusicChooser(R.id.nav_queue), 200);
                    break;
                case R.id.nav_library:
                    new Handler().postDelayed(() -> setMusicChooser(R.id.nav_library), 200);
                    break;
                case R.id.nav_folders:
                    new Handler().postDelayed(() -> setMusicChooser(R.id.nav_folders), 200);
                    break;
                case R.id.nav_eq:
                    new Handler().postDelayed(() -> setMusicChooser(R.id.nav_eq), 200);
                case R.id.buy_pro:
                    new Handler().postDelayed(() -> startActivity(new Intent(MainActivity.this, PurchaseActivity.class)), 200);
                    break;
                case R.id.action_scan:
                    new Handler().postDelayed(() -> {
                        ScanMediaFolderChooserDialog dialog = ScanMediaFolderChooserDialog.create();
                        dialog.show(getSupportFragmentManager(), "SCAN_MEDIA_FOLDER_CHOOSER");
                    }, 200);
                    break;
                case R.id.nav_settings:
                    new Handler().postDelayed(() -> startActivity(new Intent(MainActivity.this, SettingsActivity.class)), 200);
                    break;
            }
            return true;
        });
    }

    private void checkSetUpPro() {
        navigationView.getMenu().setGroupVisible(R.id.navigation_drawer_menu_category_buy_pro, !App.isProVersion());
    }

    private void setUpDrawerLayout() {
        setUpNavigationView();
    }



    @Override
    public void onServiceConnected() {
        super.onServiceConnected();
        handlePlaybackIntent(getIntent());
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == android.R.id.home) {
            if (drawerLayout.isDrawerOpen(navigationView)) {
                drawerLayout.closeDrawer(navigationView);
            } else {
                drawerLayout.openDrawer(navigationView);
            }
            return true;
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public boolean handleBackPress() {
        if (drawerLayout.isDrawerOpen(navigationView)) {
            drawerLayout.closeDrawers();
            return true;
        }
        return super.handleBackPress() || (currentFragment != null && currentFragment.handleBackPress());
    }

    private void handlePlaybackIntent(@Nullable Intent intent) {
        if (intent == null) {
            return;
        }

        Uri uri = intent.getData();
        String mimeType = intent.getType();
        boolean handled = false;

        if (intent.getAction() != null && intent.getAction().equals(MediaStore.INTENT_ACTION_MEDIA_PLAY_FROM_SEARCH)) {
            final List<Song> songs = SearchQueryHelper.getSongs(this, intent.getExtras());
            if (MusicPlayerRemote.getShuffleMode() == MusicService.SHUFFLE_MODE_SHUFFLE) {
                MusicPlayerRemote.openAndShuffleQueue(songs, true);
            } else {
                MusicPlayerRemote.openQueue(songs, 0, true);
            }
            handled = true;
        }

        if (uri != null && uri.toString().length() > 0) {
            MusicPlayerRemote.playFromUri(uri);
            handled = true;
        } else if (MediaStore.Audio.Playlists.CONTENT_TYPE.equals(mimeType)) {
            final int id = (int) parseIdFromIntent(intent, "playlistId", "playlist");
            if (id >= 0) {
                int position = intent.getIntExtra("position", 0);
                List<Song> songs = new ArrayList<>(PlaylistSongLoader.getPlaylistSongList(this, id));
                MusicPlayerRemote.openQueue(songs, position, true);
                handled = true;
            }
        } else if (MediaStore.Audio.Albums.CONTENT_TYPE.equals(mimeType)) {
            final int id = (int) parseIdFromIntent(intent, "albumId", "album");
            if (id >= 0) {
                int position = intent.getIntExtra("position", 0);
                MusicPlayerRemote.openQueue(AlbumLoader.getAlbum(this, id).songs, position, true);
                handled = true;
            }
        } else if (MediaStore.Audio.Artists.CONTENT_TYPE.equals(mimeType)) {
            final int id = (int) parseIdFromIntent(intent, "artistId", "artist");
            if (id >= 0) {
                int position = intent.getIntExtra("position", 0);
                MusicPlayerRemote.openQueue(ArtistLoader.getArtist(this, id).getSongs(), position, true);
                handled = true;
            }
        }
        if (handled) {
            setIntent(new Intent());
        }
    }

    private long parseIdFromIntent(@NonNull Intent intent, String longKey,
                                   String stringKey) {
        long id = intent.getLongExtra(longKey, -1);
        if (id < 0) {
            String idString = intent.getStringExtra(stringKey);
            if (idString != null) {
                try {
                    id = Long.parseLong(idString);
                } catch (NumberFormatException e) {
                    Log.e(TAG, e.getMessage());
                }
            }
        }
        return id;
    }

    private boolean checkShowIntro() {
        if (!PreferenceUtil.introShown()) {
            PreferenceUtil.setIntroShown();
            ChangelogDialog.setChangelogRead(this);
            blockRequestPermissions = true;
            new Handler().postDelayed(() -> startActivityForResult(new Intent(MainActivity.this, AppIntroActivity.class), APP_INTRO_REQUEST), 50);
            return true;
        }
        return false;
    }

    private void showChangelog() {
        try {
            PackageInfo pInfo = getPackageManager().getPackageInfo(getPackageName(), 0);
            int currentVersion = pInfo.versionCode;
            if (currentVersion != PreferenceUtil.getLastChangelogVersion()) {
                ChangelogDialog.create().show(getSupportFragmentManager(), "CHANGE_LOG_DIALOG");
            }
        } catch (PackageManager.NameNotFoundException e) {
            e.printStackTrace();
        }
    }


    public interface MainActivityFragmentCallbacks {
        boolean handleBackPress();
    }
}

File: app/src/main/java/com/o4x/musical/ui/activities/MainActivity.kt
package com.o4x.musical.ui.activities


import android.annotation.SuppressLint
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Bundle
import android.provider.MediaStore
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import androidx.appcompat.app.ActionBarDrawerToggle
import androidx.core.view.isVisible
import androidx.drawerlayout.widget.DrawerLayout
import androidx.lifecycle.lifecycleScope
import androidx.navigation.NavController
import code.name.monkey.appthemehelper.extensions.surfaceColor
import com.google.android.material.appbar.AppBarLayout
import com.google.android.material.appbar.MaterialToolbar
import com.google.android.material.tabs.TabLayout
import com.o4x.musical.App
import com.o4x.musical.R
import com.o4x.musical.databinding.*
import com.o4x.musical.extensions.findNavController
import com.o4x.musical.helper.MusicPlayerRemote
import com.o4x.musical.helper.SearchQueryHelper.getSongs
import com.o4x.musical.interfaces.CabHolder
import com.o4x.musical.model.Song
import com.o4x.musical.prefs.AppPref
import com.o4x.musical.repository.PlaylistSongsLoader
import com.o4x.musical.service.MusicService
import com.o4x.musical.ui.activities.base.AbsMusicPanelActivity
import com.o4x.musical.ui.activities.intro.PermissionActivity
import com.o4x.musical.views.BreadCrumbLayout
import kotlinx.android.synthetic.main.activity_main.*
import kotlinx.android.synthetic.main.activity_main_drawer_layout.*
import kotlinx.android.synthetic.main.search_bar.*
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import org.koin.android.ext.android.get

class MainActivity : AbsMusicPanelActivity(), CabHolder {

    companion object {
        val TAG: String = MainActivity::class.java.simpleName

        private val popupAbleFragments: Array<Int> =
            arrayOf(R.id.search, R.id.detail_playlist, R.id.detail_genre)
    }

    lateinit var navController: NavController
    lateinit var toggle: ActionBarDrawerToggle
    lateinit var navigationHeaderBinding: ActivityMainNavHeaderBinding

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)


        if (!hasPermissions()) {
            val myIntent = Intent(
                this,
                PermissionActivity::class.java
            )

            this.startActivity(myIntent)
        }

        AppPref.registerOnSharedPreferenceChangedListener(this)

        setDrawUnderStatusBar()
        setNavigationBarColorAuto()
        setStatusBarColorAuto()
        setLightNavigationBar(true)

        setupNavController()
        setupToolbar()
        setUpNavigationView()
    }

    override fun onResume() {
        super.onResume()
        checkVersion()
    }

    override fun onDestroy() {
        AppPref.unregisterOnSharedPreferenceChangedListener(this)
        super.onDestroy()
    }

    override fun onSupportNavigateUp(): Boolean {
        return navController.navigateUp() || super.onSupportNavigateUp()
    }

    override fun createContentView(): View {
        @SuppressLint("InflateParams") val contentView =
            layoutInflater.inflate(R.layout.activity_main_drawer_layout, null)
        val drawerContent = contentView.findViewById<ViewGroup>(R.id.drawer_content_container)
        drawerContent.addView(wrapSlidingMusicPanel(R.layout.activity_main))

        return contentView
    }

    private fun setupToolbar() {
        toolbar.setNavigationIcon(R.drawable.ic_menu)
        setSupportActionBar(toolbar)
    }

    fun setMusicChooser(id: Int) {

        if (id == navigation_view.checkedItem?.itemId ) {
            return
        }

        when (id) {
            R.id.nav_home -> navController.popBackStack()
            R.id.nav_queue -> navController.navigate(R.id.action_to_queue)
            R.id.nav_library -> navController.navigate(R.id.action_to_library)
            R.id.nav_folders -> navController.navigate(R.id.action_to_folders)
            R.id.nav_eq -> navController.navigate(R.id.action_to_equalizer)
            R.id.nav_timer -> navController.navigate(R.id.action_to_timer)
        }
    }

    fun openSearch() {
        navController.navigate(R.id.action_to_search)
    }

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences?, key: String?) {
        super.onSharedPreferenceChanged(sharedPreferences, key)
        if (key == AppPref.IS_CLEAN) {
            checkVersion()
        }
    }

    private fun checkVersion() {
        navigationHeaderBinding.apply {
            if (App.isCleanVersion()) {
                proIcon.isVisible = false
                proNext.isVisible = false
                proSummary.isVisible = false
                proTitle.text = resources.getString(R.string.musical)
                banner.setOnClickListener {  }
            } else {
                proIcon.isVisible = true
                proNext.isVisible = true
                proSummary.isVisible = true
                proTitle.text = resources.getString(R.string.musical_clean)
                banner.setOnClickListener { navController.navigate(R.id.buy_from_google_play) }
            }
        }
    }

    private fun setUpNavigationView() {
        toggle = object : ActionBarDrawerToggle(
            this,
            drawer_layout,
            R.string.open,
            R.string.close
        ) {}
        drawer_layout.addDrawerListener(toggle)
        toggle.syncState()

        navigation_view.setBackgroundColor(surfaceColor())
        navigationHeaderBinding = ActivityMainNavHeaderBinding.inflate(layoutInflater)
        navigation_view.addHeaderView(navigationHeaderBinding.root)

        var lastItem: MenuItem? = null
        drawer_layout.addDrawerListener(object : DrawerLayout.DrawerListener {

            override fun onDrawerSlide(drawerView: View, slideOffset: Float) {}
            override fun onDrawerStateChanged(newState: Int) {}
            override fun onDrawerOpened(drawerView: View) {}

            override fun onDrawerClosed(drawerView: View) {

                when (lastItem?.itemId) {
                    R.id.nav_home -> setMusicChooser(R.id.nav_home)
                    R.id.nav_queue -> setMusicChooser(R.id.nav_queue)
                    R.id.nav_library -> setMusicChooser(R.id.nav_library)
                    R.id.nav_folders -> setMusicChooser(R.id.nav_folders)
                    R.id.nav_eq -> setMusicChooser(R.id.nav_eq)
                    R.id.nav_timer -> setMusicChooser(R.id.nav_timer)
                    R.id.nav_settings -> navController.navigate(R.id.settings)
                }

            }
        })

        navigation_view.setNavigationItemSelectedListener { menuItem: MenuItem ->
            lastItem = menuItem
            drawer_layout.closeDrawers()

            false
        }
    }

    fun setDrawerEnabled(enabled: Boolean) {
        val lockMode: Int =
            if (enabled) DrawerLayout.LOCK_MODE_UNLOCKED else DrawerLayout.LOCK_MODE_LOCKED_CLOSED
        drawer_layout.setDrawerLockMode(lockMode)
        toggle.isDrawerIndicatorEnabled = enabled
        toggle.syncState()
    }

    override fun onServiceConnected() {
        super.onServiceConnected()
        handlePlaybackIntent(intent)
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if (item.itemId == android.R.id.home) {

            for (id in popupAbleFragments) {
                if (navController.currentDestination?.id == id)
                    return false
            }

            if (drawer_layout.isDrawerOpen(navigation_view)) {
                drawer_layout.closeDrawer(navigation_view)
            } else {
                drawer_layout.openDrawer(navigation_view)
            }
            return true
        }
        return super.onOptionsItemSelected(item)
    }

    override fun handleBackPress(): Boolean {
        if (drawer_layout.isDrawerOpen(navigation_view)) {
            drawer_layout.closeDrawers()
            return true
        }
        backPressCallbacks.forEach {
            if (it.handleBackPress()) {
                return true
            }
        }
        return super.handleBackPress() || navController.popBackStack()
    }

    private fun handlePlaybackIntent(intent: Intent) {
        lifecycleScope.launch(Dispatchers.IO) {
            val uri: Uri? = intent.data
            val mimeType: String? = intent.type
            var handled = false
            if (intent.action != null &&
                intent.action == MediaStore.INTENT_ACTION_MEDIA_PLAY_FROM_SEARCH
            ) {
                val songs: List<Song> = getSongs(intent.extras!!)
                if (MusicPlayerRemote.shuffleMode == MusicService.SHUFFLE_MODE_SHUFFLE) {
                    MusicPlayerRemote.openAndShuffleQueue(songs, true)
                } else {
                    MusicPlayerRemote.openQueue(songs, 0, true)
                }
                handled = true
            }
            if (uri != null && uri.toString().isNotEmpty()) {
                MusicPlayerRemote.playFromUri(uri)
                handled = true
            } else if (MediaStore.Audio.Playlists.CONTENT_TYPE == mimeType) {
                val id = parseLongFromIntent(intent, "playlistId", "playlist")
                if (id >= 0L) {
                    val position: Int = intent.getIntExtra("position", 0)
                    val songs: List<Song> = PlaylistSongsLoader.getPlaylistSongList(get(), id)
                    MusicPlayerRemote.openQueue(songs, position, true)
                    handled = true
                }
            } else if (MediaStore.Audio.Albums.CONTENT_TYPE == mimeType) {
                val id = parseLongFromIntent(intent, "albumId", "album")
                if (id >= 0L) {
                    val position: Int = intent.getIntExtra("position", 0)
                    val songs = libraryViewModel.albumById(id).songs
                    MusicPlayerRemote.openQueue(
                        songs,
                        position,
                        true
                    )
                    handled = true
                }
            } else if (MediaStore.Audio.Artists.CONTENT_TYPE == mimeType) {
                val id = parseLongFromIntent(intent, "artistId", "artist")
                if (id >= 0L) {
                    val position: Int = intent.getIntExtra("position", 0)
                    val songs: List<Song> = libraryViewModel.artistById(id).songs
                    MusicPlayerRemote.openQueue(
                        songs,
                        position,
                        true
                    )
                    handled = true
                }
            }
            if (handled) {
                setIntent(Intent())
            }
        }

    }

    private fun parseLongFromIntent(
        intent: Intent, longKey: String,
        stringKey: String
    ): Long {
        var id = intent.getLongExtra(longKey, -1)
        if (id < 0) {
            val idString = intent.getStringExtra(stringKey)
            if (idString != null) {
                try {
                    id = idString.toLong()
                } catch (e: NumberFormatException) {
                    println(e.message)
                }
            }
        }
        return id
    }

    val backPressCallbacks = mutableListOf<MainActivityFragmentCallbacks>()

    interface MainActivityFragmentCallbacks {
        fun handleBackPress(): Boolean
    }

    private fun setupNavController() {
        navController = findNavController(R.id.fragment_container)

        navController.addOnDestinationChangedListener { _, destination, _ ->
            when (destination.id) {
                R.id.home -> navigation_view.setCheckedItem(R.id.nav_home)
                R.id.queue -> navigation_view.setCheckedItem(R.id.nav_queue)
                R.id.library -> navigation_view.setCheckedItem(R.id.nav_library)
                R.id.folders -> navigation_view.setCheckedItem(R.id.nav_folders)
                R.id.equalizer -> navigation_view.setCheckedItem(R.id.nav_eq)
                R.id.timer -> navigation_view.setCheckedItem(R.id.nav_timer)
            }
        }
    }

    val appbar: AppBarLayout
        get() = main_appbar

    val toolbar: MaterialToolbar
        get() = main_toolbar

    val search: LinearLayout
        get() = main_search

    val tabs: TabLayout
        get() = main_tabs

    val bread_crumbs: BreadCrumbLayout
        get() = main_bread_crumbs
}